framework/SupervisedLearning/SyntheticHistory.py

```diff
@@ -0,0 +1,202 @@
+    # TODO assumption: only one training signal
+    pivots = targetVals[0, :, pivotIndex]
+    self.pivotParameterValues = pivots[:] # TODO any way to avoid storing these?
+    residual = targetVals[:, :, :] # deep-ish copy, so we don't mod originals
+    for algo in self.tsaAlgorithms:
+      targets = algo.target
+      indices = tuple(self.target.index(t) for t in targets)
+      signal = residual[0, :, indices].T # using tuple "indices" transposes, so transpose back
+      params = algo.characterize(signal, pivots, targets)
+      # store characteristics
+      self.trainedParams[algo] = params
+      # obtain residual; the part of the signal not characterized by this algo
+      algoResidual = algo.getResidual(signal, params, pivots, None) # TODO randomEngine
+      residual[0, :, indices] = algoResidual.T # transpose, again because of indices
+      # TODO meta store signal, residual?
+
+  def __evaluateLocal__(self, featureVals):
+    """
+      @ In, featureVals, float, a scalar feature value is passed as scaling factor
+      @ Out, rlz, dict, realization dictionary of values for each target
+    """
+    pivots = self.pivotParameterValues
+    result = np.zeros((self.pivotParameterValues.size, len(self.target) - 1)) # -1 is pivot
+    for algo in self.tsaAlgorithms:
+      targets = algo.target
+      indices = tuple(self.target.index(t) for t in targets)
+      params = self.trainedParams[algo]
+      signal = algo.generate(params, pivots, None) # TODO random engine for third argument
+      result[:, indices] = signal
+    # RAVEN realization construction
+    rlz = dict((target, signal[:, t]) for t, target in enumerate(self.target) if target != self.pivotParameterID)
+    rlz[self.pivotParameterID] = self.pivotParameterValues
+    return rlz
+
+  def writePointwiseData(self, writeTo):
+    """
+      Writes pointwise data about this ROM to the data object.
+      @ In, writeTo, DataObject, data structure into which data should be written
+      @ Out, None
+    """
+    pass # TODO
+
+  def writeXML(self, writeTo, targets=None, skip=None):
+    """
+      Allows the SVE to put whatever it wants into an XML to print to file.
+      Overload in subclasses.
+      @ In, writeTo, xmlUtils.StaticXmlElement, entity to write to
+      @ In, targets, list, optional, unused (kept for compatability)
+      @ In, skip, list, optional, unused (kept for compatability)
+      @ Out, None
+    """
+    root = writeTo.getRoot()
+    for algo in self.tsaAlgorithms:
+      algoNode = xmlUtils.newNode(algo.name)
+      algo.writeXML(algoNode, self.trainedParams[algo])
+      root.append(algoNode)
+
+  ### Segmenting and Clustering ###
+  def isClusterable(self):
+    """
+      Allows ROM to declare whether it has methods for clustring. Default is no.
+      @ In, None
+      @ Out, isClusterable, bool, if True then has clustering mechanics.
+    """
+    # clustering methods have been added
+    return False # TODO
+
+  ### ESSENTIALLY UNUSED ###
```

Comment:
What does essentially unused mean? Maybe a block comment explaining that? Your call.

Reply:
It means the original SVL designer put in a bunch of abstract classes that only a couple of the SVLs use, but my request to remove the abstract methods has been denied, so we include these instead. I think this is explained in one of the SVL classes, maybe.